FileManager/BackgroundFileSystem.cs

```diff
@@ -0,0 +1,117 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+namespace FileManager
+{
+    class BackgroundFileSystem
+    {
+        public string RootDirectory { get; private set; }
+        public string SearchPattern { get; private set; }
+        public SearchOption SearchOption { get; private set; }
+
+        private FileSystemWatcher fileSystemWatcher { get; set; }
+        private BlockingCollection<FileSystemEventArgs> directoryChangesEvents { get; set; }
+        private Task backgroundScanner { get; set; }
+        private List<string> fsCache { get; set; }
+
+        public string FindFile(string regexPattern, RegexOptions options)
+        {
+            lock (fsCache)
+            {
+                return fsCache.FirstOrDefault(s => Regex.IsMatch(s, regexPattern, options));
+            }
+        }
+
+        public void Init(string rootDirectory, string searchPattern, SearchOption searchOptions)
+        {
+            RootDirectory = rootDirectory;
+            SearchPattern = searchPattern;
+            SearchOption = searchOptions;
+
+            //Calling CompleteAdding() will cause background scanner to terminate.
+            directoryChangesEvents?.CompleteAdding();
+            fsCache?.Clear();
+            directoryChangesEvents?.Dispose();
+            fileSystemWatcher?.Dispose();
+
+            fsCache = Directory.EnumerateFiles(RootDirectory, SearchPattern, SearchOption).ToList();
+
+            directoryChangesEvents = new BlockingCollection<FileSystemEventArgs>();
+            fileSystemWatcher = new FileSystemWatcher(RootDirectory);
+            fileSystemWatcher.Created += FileSystemWatcher_Changed;
+            fileSystemWatcher.Deleted += FileSystemWatcher_Changed;
+            fileSystemWatcher.Renamed += FileSystemWatcher_Changed;
```

Comment:
Last time I worked with system file watcher, these were fired a lot, esp anywhere with a lot of changes. Instead of adding to the queue, I suggest add if not already present. Else the queue can contain a lot of duplicates and waste a lot of time scanning the same dir back to back

Reply:
What I observed playing around with it is that there are only redundant firings if you subscribe to FileSystemWatcher.Changed, because it gets fired at the same time as Created, Deleted, and Renamed. Buf if you don't subscribe to created, then each of the other 3 events fire only once per change as expected. Renaming a file or directiory fires once. Deleting a directory, including all its contents, fires Deleted only once. Adding a directory, including one with contents, fires Created only once.

Reply:
Cool. Maybe that's what I had been doing wrong at the time